### PR TITLE
Add validation tests of CopyTextureToTexture - Part I

### PR DESCRIPTION
**Pull request by Jiawei-Shao**

### CODE REVIEW

File: src/webgpu/api/validation/copyTextureToTexture.spec.ts [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+      srcCopyMipLevel: 0,
+      dstCopyMipLevel: 3,
+      _isSuccess: false
+    },
+    {
+      srcMipLevelCount: 3,
+      dstMipLevelCount: 3,
+      srcCopyMipLevel: 0,
+      dstCopyMipLevel: 4,
+      _isSuccess: false
```
Comment (Kangz) [inline src/webgpu/api/validation/copyTextureToTexture.spec.ts:134]: nit: _isSuccess could be computed programmatically.

File: src/webgpu/api/validation/copyTextureToTexture.spec.ts [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+      _isSuccess: true
+    },
+    {
+      srcMipLevelCount: 3,
+      dstMipLevelCount: 3,
+      srcCopyMipLevel: 3,
+      dstCopyMipLevel: 0,
+      _isSuccess: false
+    },
+    {
```
Comment (Kangz) [inline src/webgpu/api/validation/copyTextureToTexture.spec.ts:108]: nit: this case seems covered by the one above already

File: src/webgpu/api/validation/copyTextureToTexture.spec.ts [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+      srcMipLevelCount: 3,
+      dstMipLevelCount: 3,
+      srcCopyMipLevel: 0,
+      dstCopyMipLevel: 3,
+      _isSuccess: false
+    },
+    {
+      srcMipLevelCount: 3,
```
Comment (Kangz) [inline src/webgpu/api/validation/copyTextureToTexture.spec.ts:130]: ditto

File: src/webgpu/api/validation/copyTextureToTexture.spec.ts [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+      { texture: srcTexture },
+      { texture: dstTexture },
+      { width: 1, height: 1, depth: 1 },
+      isSuccess
+    );
+  });
+
+g.test('sample_count')
+  .params([
+    { srcSamplecount: 1, dstSampleCount: 1, _isSuccess: true },
```
Comment (Kangz) [inline src/webgpu/api/validation/copyTextureToTexture.spec.ts:201]: ditto: compute isSuccess programmatially

File: src/webgpu/api/validation/copyTextureToTexture.spec.ts [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+      _isSuccess
+    );
+  });
+
+g.test('multisampled_copy_restrictions')
+  .params(
+    params()
+      .combine(
+        poptions('srcCopyOrigin', [
+          { x: 0, y: 0, z: 0 },
```
Comment (Kangz) [inline src/webgpu/api/validation/copyTextureToTexture.spec.ts:235]: nit: also have a test for origin = 0, 0, 0 but the width / height / depth not being the full subresource size (and also at a miplevel to check the miplevel size is correctly used)

Comment (Jiawei-Shao) [reply src/webgpu/api/validation/copyTextureToTexture.spec.ts:234]: The mipmap level count of multisampled textures must be 1 so we shouldn't consider mipLevel > 0 here.

**Comment by kainino0x** [comment]
FYI just in case you missed it - you can use `npx grunt check` and `npx grunt fix` locally for the formatting issues.

File: src/webgpu/api/validation/copyTextureToTexture.spec.ts [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    { texture: errorTexture },
+    { width: 1, height: 1, depth: 1 },
+    false
+  );
+});
+
+g.test('mipmap_level')
+  .params([
+    {
+      srcMipLevelCount: 1,
```
Comment (kainino0x) [inline src/webgpu/api/validation/copyTextureToTexture.spec.ts:67]: uNit: Consider using shorter names here so these can fit on one line each (would be easier to read). Maybe change `MipLevel` to just `Level`.

**Comment by Jiawei-Shao** [comment]
> FYI just in case you missed it - you can use `npx grunt check` and `npx grunt fix` locally for the formatting issues.

@kainino0x  Oh sorry I missed it before. I'll pay attention to it.

**Review by kainino0x — APPROVED** [review]
Nice and easy to understand tests. LGTM

File: src/webgpu/api/validation/copyTextureToTexture.spec.ts (hunk 6 above)
Comment (Jiawei-Shao) [reply src/webgpu/api/validation/copyTextureToTexture.spec.ts]: Done

File: src/webgpu/api/validation/copyTextureToTexture.spec.ts (hunk 5 above)
Comment (Jiawei-Shao) [reply src/webgpu/api/validation/copyTextureToTexture.spec.ts:167]: Done

**Review by Kangz — APPROVED** [review]

**Pull request merged**